File: Cpp/Maximal Square.cpp
// 20ms, 80.00%; 11.2MB, 51.85%
class Solution
{
public:
    int maximalSquare(const vector<vector<char>> &matrix)
    {
        if (matrix.empty()) return 0;
        int row = matrix.size(), col = matrix[0].size();
        vector<vector<int>> dp(row, vector<int>(col, 0));
        int length = helper(matrix, dp);
        for (int i = 1; i < row; ++i) {
            for (int j = 1; j < col; ++j) {
                if (matrix[i][j] == '0') continue;
                dp[i][j] = 1 + min({dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]});
                length = max(length, dp[i][j]);
            }
        }
        return length * length;
    }

private:
    int helper(const vector<vector<char>> &matrix, vector<vector<int>> &dp)
    {
        int flag = 0;
        int row = matrix.size(), col = matrix[0].size();
        for (int i = 0; i < row; ++i)
            if (matrix[i][0] == '1') dp[i][0] = flag = 1;
        for (int j = 1; j < col; ++j)
            if (matrix[0][j] == '1') dp[0][j] = flag = 1;
        return flag;
    }
};

File: Python3/Maximal Square.py
# 188ms, 91.51%; 13.7MB, 100.00%
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix: return 0
        row, col, length = len(matrix), len(matrix[0]), self.__helper(matrix)
        dp1, dp2 = [1 if x == '1' else 0 for x in matrix[0]], [0] * col
        for i in range(1, row):
            dp2[0] = 1 if matrix[i][0] == '1' else 0
            for j in range(1, col):
                if matrix[i][j] == '0': continue
                dp2[j] = 1 + min(dp1[j - 1], dp1[j], dp2[j - 1])
                length = max(length, dp2[j])
            dp1, dp2 = dp2, [0] * col
        return length**2

    def __helper(self, matrix: List[List[str]]) -> int:
        for x in matrix[0]:
            if x == '1': return 1
        for i in range(1, len(matrix)):
            if matrix[i][0] == '1': return 1
        return 0
